### ISSUE TITLE
Function execute_command don't respect document workspace

### ISSUE THREAD

**Issue by mireq** [python-lsp/python-lsp-server#335]
Function `execute_command` calls `_hook` without `doc_uri`. Variable `workspace` is determined in `_hook` only as fallback for first opened workspace. Now refactoring tool like ` pylsp-rope` don't have correct `workspace`.

I don't know how to solve this problem. Maybe, it should exist standard way to pass `doc_uri` from `pylsp_code_actions` to `execute_command`.

**Comment by Shabbyshambles** [comment]
Ok, I see what you mean, In Python_lsp.py, the "_hook" function sets the "doc_uri" value to None, and then uses that value to get the workspace variable. So the variable of "workspace" is set to None, so the doc variable is set to None.
And since the Workspace is None, you won't have access to key elements of pylsp-rope.
If I understand this correctly, what we are looking at is an integration issue with the refactoring software of rope and the execute_command function. 
Can you enter the doc_uri manually using the "command" or "arguments" parameters in the execute_command function? or maybe there exists a way to get rope to refactor automatically for a given workspace eliminating the need for the execute_command to manually refactor?
Or maybe we could bundle the pylsp_code_actions and execute_command hook-names into one list shared between all functions that use the two? This is a Gordian Knot of an issue

**Comment by mireq** [comment]
Refactoring plugin `pylsp-rope` sends document uri as `document_uri` parameter. I have modified my `_hook` function with this ugly hack:

```python
def _hook(self, hook_name, doc_uri=None, **kwargs):
    """Calls hook_name and returns a list of results from all registered handlers"""
    if doc_uri is None and kwargs.get('arguments') and 'document_uri' in kwargs['arguments'][0]:
        doc_uri = kwargs['arguments'][0]['document_uri']
    workspace = self._match_uri_to_workspace(doc_uri)
    doc = workspace.get_document(doc_uri) if doc_uri else None
    hook_handlers = self.config.plugin_manager.subset_hook_caller(hook_name, self.config.disabled_plugins)
    return hook_handlers(config=self.config, workspace=workspace, document=doc, **kwargs)
```

It works, but ... there is no standard way to exchange `doc_uri` between refactoring plugin like `pylsp-rope`  and `python-lsp-server`. In future, `pylsp-rope` can change argument value and my hack stops working.

**Comment by Shabbyshambles** [comment]
Yeah thats a bit messy, but if it works hard to knock it. The real solution is probably build out a function that onboards the doc_uri from pylsp-rope, build out a function in pylsp-rope that sends doc_uri to the server, and then have all functions that want to exchange doc-uri do so through those two functions but that is probably way too cumbersome to be actually implemented. 
Or we reach out to the devs of pylsp-rope and say hey, please please please don't change the argument value, and if you do let us know so we can update this snippet

**Comment by ccordoba12** [comment]
> Function execute_command calls _hook without doc_uri

Is it not enough to extend `execute_command` here so that it receives `doc_uri`?

**Comment by mireq** [comment]
@ccordoba12 action execute_command is defined by [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#executeCommandParams). It passes only two arguments - command name and user defined array of arguments.

**Comment by ccordoba12** [comment]
> It passes only two arguments - command name and user defined array of arguments.

Ok, I see. Then, what if you send `doc_uri` in the array of arguments, we detect if it corresponds to an actual `doc_uri` in `execute_command`, and if that's the case pass it to `_hook`?

**Comment by ccordoba12** [comment]
Since `arguments` is the list of arguments for `command`, then I think you could send `doc_uri` in a dict.

**Comment by mireq** [comment]
@ccordoba12 `pylsp-rope` already sends `doc_uri` as part of arguments. Problem is, that `pylsp-rope` cannot get workspace by `doc_uri` (its responsibility of `python-lsp-server`) and `python-lsp-server` cannot get `doc_uri` from arguments because format is defined by `pylsp-rope` and it can be changed in future.

**Comment by Shabbyshambles** [comment]
Seems to me our options are to either maintain the code snippet based on changes made in pylsp-rope, or get in touch with the team working on that project and try to get them to preserve the structure they are using to pass the argument.

**Comment by mireq** [comment]
@Shabbyshambles or expose something like `get_workspace_by_uri` as public API and call it from `pylsp-rope`.

**Comment by ccordoba12** [comment]
> @Shabbyshambles or expose something like get_workspace_by_uri as public API and call it from pylsp-rope

Yeah, that's what we should do on our side to help you with this.

@mireq or @Shabbyshambles, would you like to submit a PR for that?

**Comment by Shabbyshambles** [comment]
I'm tied up until Tuesday but I can take a crack at it then

**Comment by ccordoba12** [comment]
That's ok and thanks for your help with that!
